File: backend/src/redforge/api/v1/investigations.py
"""Unified Threat Investigation REST API — M21.

All routes are tenant-scoped; organization_id is derived from JWT.

Permission gates:
  INVESTIGATIONS_READ   — all read endpoints
  INVESTIGATIONS_MANAGE — acknowledge / start-investigation / resolve
"""

from __future__ import annotations

from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from redforge.api.security import TenantContext, require_permission
from redforge.domain.identity.value_objects import Permission
from redforge.domain.investigations.exceptions import (
    InvalidStatusTransitionError,
    InvestigationNotFoundError,
)

router = APIRouter(prefix="/investigations")


# ── Request models ────────────────────────────────────────────────────────────


class AcknowledgeRequest(BaseModel):
    pass  # no body required


class StartInvestigationRequest(BaseModel):
    pass  # no body required


class ResolveRequest(BaseModel):
    resolution_reason: str = Field(..., min_length=1, max_length=60)
    notes: str = Field("", max_length=2000)


# ── Dependency helpers ────────────────────────────────────────────────────────


def _get_session_factory() -> Any:
    from redforge.api.dependencies import _session_factory
    return _session_factory()


async def _investigation_svc(
    session_factory: Any = Depends(_get_session_factory),
) -> Any:
    from redforge.application.investigations.case_service import InvestigationCaseService
    from redforge.application.security_graph.projector import SecurityGraphProjector
    from redforge.infrastructure.audit.logger import StructlogAuditLog
    from redforge.infrastructure.database.repositories.investigations.case_repository import (
        SqlAlchemyEvidenceLinkRepository,
        SqlAlchemyInvestigationEventRepository,
        SqlAlchemyInvestigationRepository,
    )
    from redforge.infrastructure.database.repositories.security_graph_repository import (
        SecurityGraphRepository,
    )

    async with session_factory() as session, session.begin():
        case_repo = SqlAlchemyInvestigationRepository(session)
        evidence_repo = SqlAlchemyEvidenceLinkRepository(session)
        event_repo = SqlAlchemyInvestigationEventRepository(session)
        audit_log = StructlogAuditLog()
        graph_repo = SecurityGraphRepository(session)
        graph_projector = SecurityGraphProjector(graph_repo)
        yield InvestigationCaseService(
            session, case_repo, evidence_repo, event_repo, audit_log, graph_projector
        )


async def _repos(session_factory: Any = Depends(_get_session_factory)) -> Any:
    from redforge.infrastructure.database.repositories.investigations.case_repository import (
        SqlAlchemyEvidenceLinkRepository,
        SqlAlchemyInvestigationEventRepository,
        SqlAlchemyInvestigationRepository,
    )

    async with session_factory() as session, session.begin():
        yield (
            SqlAlchemyInvestigationRepository(session),
            SqlAlchemyEvidenceLinkRepository(session),
            SqlAlchemyInvestigationEventRepository(session),
        )


# ── Formatters ────────────────────────────────────────────────────────────────


def _fmt_case(case: Any) -> dict[str, Any]:
    return {
        "id": case.id,
        "organization_id": case.organization_id,
        "title": case.title,
        "summary": case.summary,
        "status": case.status,
        "severity": case.severity,
        "confidence": case.confidence,
        "source_domains": case.source_domains or [],
        "involved_entities": case.involved_entities or [],
        "evidence_count": case.evidence_count,
        "first_observed_at": case.first_observed_at.isoformat(),
        "last_observed_at": case.last_observed_at.isoformat(),
        "opened_at": case.opened_at.isoformat(),
        "acknowledged_at": case.acknowledged_at.isoformat() if case.acknowledged_at else None,
        "investigating_at": case.investigating_at.isoformat() if case.investigating_at else None,
        "resolved_at": case.resolved_at.isoformat() if case.resolved_at else None,
        "resolution_reason": case.resolution_reason,
        "resolution_notes": case.resolution_notes,
        "version": case.version,
        "created_at": case.created_at.isoformat(),
        "updated_at": case.updated_at.isoformat(),
    }


def _fmt_evidence(link: Any) -> dict[str, Any]:
    return {
        "id": link.id,
        "case_id": link.case_id,
        "source_domain": link.source_domain,
        "source_entity_type": link.source_entity_type,
        "source_entity_id": link.source_entity_id,
        "event_type": link.event_type,
        "severity": link.severity,
        "observed_at": link.observed_at.isoformat(),
        "evidence_snapshot": link.evidence_snapshot,
        "correlation_reason": link.correlation_reason,
        "relationship_type": link.relationship_type,
        "observability": link.observability,
        "dedup_key": link.dedup_key,
        "created_at": link.created_at.isoformat(),
    }


def _fmt_event(ev: Any) -> dict[str, Any]:
    return {
        "id": ev.id,
        "event_id": ev.event_id,
        "case_id": ev.case_id,
        "event_type": ev.event_type,
        "detail": ev.detail,
        "actor_user_id": ev.actor_user_id,
        "occurred_at": ev.occurred_at.isoformat(),
    }


# ── Endpoints ─────────────────────────────────────────────────────────────────


@router.get("/posture")
async def get_investigation_posture(
    tenant: Annotated[
        TenantContext, Depends(require_permission(Permission.INVESTIGATIONS_READ))
    ],
    svc: Any = Depends(_investigation_svc),
) -> dict[str, Any]:
    """Global investigation posture for the organization."""
    return await svc.get_posture(tenant.organization_id)  # type: ignore[no-any-return]


@router.get("")
async def list_investigations(
    tenant: Annotated[
        TenantContext, Depends(require_permission(Permission.INVESTIGATIONS_READ))
    ],
    status: str | None = Query(None),
    severity: str | None = Query(None),
    source_domain: str | None = Query(None),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    repos: Any = Depends(_repos),
) -> list[dict[str, Any]]:
    """List investigation cases with optional filters (paginated)."""
    case_repo, _, _ = repos
    cases = await case_repo.list_cases(
        tenant.organization_id,
        status=status,
        severity=severity,
        source_domain=source_domain,
        limit=limit,
        offset=offset,
    )
    return [_fmt_case(c) for c in cases]


@router.get("/{case_id}")
async def get_investigation(
    case_id: str,
    tenant: Annotated[
        TenantContext, Depends(require_permission(Permission.INVESTIGATIONS_READ))
    ],
    repos: Any = Depends(_repos),
) -> dict[str, Any]:
    """Get a single investigation case."""
    case_repo, evidence_repo, _ = repos
    case = await case_repo.get(tenant.organization_id, case_id)
    if case is None:
        raise HTTPException(status_code=404, detail="Investigation not found")

    domain_count = await evidence_repo.count_domains(tenant.organization_id, case_id)
    result = _fmt_case(case)
    result["domain_count"] = domain_count
    return result


@router.get("/{case_id}/timeline")
async def get_investigation_timeline(
    case_id: str,
    tenant: Annotated[
        TenantContext, Depends(require_permission(Permission.INVESTIGATIONS_READ))
    ],
    limit: int = Query(200, ge=1, le=500),
    offset: int = Query(0, ge=0),
    repos: Any = Depends(_repos),
) -> list[dict[str, Any]]:
    """Chronological timeline events for an investigation."""
    case_repo, _, event_repo = repos
    case = await case_repo.get(tenant.organization_id, case_id)
    if case is None:
        raise HTTPException(status_code=404, detail="Investigation not found")
    events = await event_repo.list_events(
        tenant.organization_id, case_id, limit=limit, offset=offset
    )
    return [_fmt_event(e) for e in events]


@router.get("/{case_id}/evidence")
async def get_investigation_evidence(
    case_id: str,
    tenant: Annotated[
        TenantContext, Depends(require_permission(Permission.INVESTIGATIONS_READ))
    ],
    source_domain: str | None = Query(None),
    limit: int = Query(200, ge=1, le=500),
    offset: int = Query(0, ge=0),
    repos: Any = Depends(_repos),
) -> list[dict[str, Any]]:
    """Source-domain evidence attached to an investigation."""
    case_repo, evidence_repo, _ = repos
    case = await case_repo.get(tenant.organization_id, case_id)
    if case is None:
        raise HTTPException(status_code=404, detail="Investigation not found")
    links = await evidence_repo.list_for_case(
        tenant.organization_id, case_id,
        source_domain=source_domain,
        limit=limit, offset=offset,
    )
    return [_fmt_evidence(lnk) for lnk in links]


@router.get("/{case_id}/graph")
async def get_investigation_graph(
    case_id: str,
    tenant: Annotated[
        TenantContext, Depends(require_permission(Permission.INVESTIGATIONS_READ))
    ],
    repos: Any = Depends(_repos),
) -> dict[str, Any]:
    """Evidence-backed investigation graph (entities + relationships).

    Only returns OBSERVED or INFERRED relationships — never fabricated.
    """
    case_repo, evidence_repo, _ = repos
    case = await case_repo.get(tenant.organization_id, case_id)
    if case is None:
        raise HTTPException(status_code=404, detail="Investigation not found")

    links = await evidence_repo.list_for_case(tenant.organization_id, case_id, limit=200)

    nodes = []
    edges = []

    # Investigation case node
    nodes.append({
        "id": case.id,
        "kind": "investigation",
        "label": case.title,
        "severity": case.severity,
        "status": case.status,
    })

    seen_entities: set[str] = set()
    for link in links:
        entity_node_id = f"{link.source_domain}:{link.source_entity_id}"
        if entity_node_id not in seen_entities:
            seen_entities.add(entity_node_id)
            nodes.append({
                "id": entity_node_id,
                "kind": link.source_entity_type,
                "label": link.source_entity_id,
                "source_domain": link.source_domain,
            })

        edges.append({
            "source": entity_node_id,
            "target": case.id,
            "kind": "correlated_with",
            "observability": link.observability,
            "reason": link.correlation_reason[:200] if link.correlation_reason else "",
        })

    # Entity nodes from involved_entities
    for entity in (case.involved_entities or []):
        ent_id = f"entity:{entity.get('type', '')}:{entity.get('id', '')}"
        if ent_id not in seen_entities:
            seen_entities.add(ent_id)
            nodes.append({
                "id": ent_id,
                "kind": entity.get("type", "unknown").lower(),
                "label": entity.get("id", ""),
            })

    return {
        "case_id": case.id,
        "nodes": nodes,
        "edges": edges,
        "ontology_version": 6,
    }


@router.post("/{case_id}/acknowledge")
async def acknowledge_investigation(
    case_id: str,
    tenant: Annotated[
        TenantContext, Depends(require_permission(Permission.INVESTIGATIONS_MANAGE))
    ],
    _body: AcknowledgeRequest = AcknowledgeRequest(),
    svc: Any = Depends(_investigation_svc),
) -> dict[str, Any]:
    """Acknowledge an open investigation case."""
    try:
        await svc.acknowledge(
            organization_id=tenant.organization_id,
            case_id=case_id,
            actor_user_id=tenant.user_id,
        )
    except InvestigationNotFoundError:
        raise HTTPException(status_code=404, detail="Investigation not found") from None
    except InvalidStatusTransitionError as e:
        raise HTTPException(status_code=422, detail=str(e)) from e
    return {"case_id": case_id, "status": "ACKNOWLEDGED"}


@router.post("/{case_id}/start-investigation")
async def start_investigation(
    case_id: str,
    tenant: Annotated[
        TenantContext, Depends(require_permission(Permission.INVESTIGATIONS_MANAGE))
    ],
    _body: StartInvestigationRequest = StartInvestigationRequest(),
    svc: Any = Depends(_investigation_svc),
) -> dict[str, Any]:
    """Move a case into active investigation."""
    try:
        await svc.start_investigation(
            organization_id=tenant.organization_id,
            case_id=case_id,
            actor_user_id=tenant.user_id,
        )
    except InvestigationNotFoundError:
        raise HTTPException(status_code=404, detail="Investigation not found") from None
    except InvalidStatusTransitionError as e:
        raise HTTPException(status_code=422, detail=str(e)) from e
    return {"case_id": case_id, "status": "INVESTIGATING"}


@router.post("/{case_id}/resolve")
async def resolve_investigation(
    case_id: str,
    body: ResolveRequest,
    tenant: Annotated[
        TenantContext, Depends(require_permission(Permission.INVESTIGATIONS_MANAGE))
    ],
    svc: Any = Depends(_investigation_svc),
) -> dict[str, Any]:
    """Resolve an investigation case with a bounded resolution reason."""
    from redforge.domain.investigations.value_objects import ResolutionReason

    try:
        ResolutionReason(body.resolution_reason)  # validate
    except ValueError:
        valid = [r.value for r in ResolutionReason]
        raise HTTPException(  # noqa: B904
            status_code=422,
            detail=f"Invalid resolution_reason. Valid values: {valid}",
        )
    try:
        await svc.resolve(
            organization_id=tenant.organization_id,
            case_id=case_id,
            actor_user_id=tenant.user_id,
            resolution_reason=body.resolution_reason,
            notes=body.notes,
        )
    except InvestigationNotFoundError:
        raise HTTPException(status_code=404, detail="Investigation not found") from None
    except InvalidStatusTransitionError as e:
        raise HTTPException(status_code=422, detail=str(e)) from e
    return {"case_id": case_id, "status": "RESOLVED"}

File: frontend/src/lib/investigations.ts
/**
 * AIVAR RedForge — Cross-Domain Security Investigation API client (M21).
 *
 * All data is derived from canonical investigation cases.
 * No synthetic or fabricated values are introduced.
 *
 * Every case has a deterministic correlation_key linking it to a specific
 * rule and entity set. Confidence is a four-level enum, not an opaque score.
 */

import { api } from "@/lib/api";

// ── Types ─────────────────────────────────────────────────────────────────────

export type InvestigationStatus = "OPEN" | "ACKNOWLEDGED" | "INVESTIGATING" | "RESOLVED";
export type InvestigationSeverity = "CRITICAL" | "HIGH" | "MEDIUM" | "LOW";
export type CorrelationConfidence = "LOW" | "MEDIUM" | "HIGH" | "VERY_HIGH";

export interface InvolvedEntity {
  type: string;
  id: string;
}

export interface InvestigationCase {
  id: string;
  organization_id: string;
  title: string;
  summary: string;
  status: InvestigationStatus;
  severity: InvestigationSeverity;
  confidence: CorrelationConfidence;
  source_domains: string[];
  involved_entities: InvolvedEntity[];
  evidence_count: number;
  first_observed_at: string;
  last_observed_at: string;
  opened_at: string;
  acknowledged_at: string | null;
  investigating_at: string | null;
  resolved_at: string | null;
  resolution_reason: string | null;
  resolution_notes: string | null;
  version: number;
  created_at: string;
  updated_at: string;
  domain_count?: number;
}

export interface InvestigationPosture {
  open_cases: number;
  acknowledged_cases: number;
  investigating_cases: number;
  resolved_cases: number;
  critical_cases: number;
  high_cases: number;
  multi_domain_cases: number;
  total_active: number;
}

export interface EvidenceLink {
  id: string;
  case_id: string;
  source_domain: string;
  source_entity_type: string;
  source_entity_id: string;
  event_type: string;
  severity: string;
  observed_at: string;
  evidence_snapshot: Record<string, unknown>;
  correlation_reason: string;
  relationship_type: string;
  observability: string;
  dedup_key: string;
  created_at: string;
}

export interface InvestigationTimelineEvent {
  id: string;
  event_id: string;
  case_id: string;
  event_type: string;
  detail: Record<string, unknown>;
  actor_user_id: string | null;
  occurred_at: string;
}

export interface InvestigationGraphNode {
  id: string;
  kind: string;
  label: string;
  severity?: string;
  status?: string;
  source_domain?: string;
}

export interface InvestigationGraphEdge {
  source: string;
  target: string;
  kind: string;
  observability: string;
  reason: string;
}

export interface InvestigationGraph {
  case_id: string;
  nodes: InvestigationGraphNode[];
  edges: InvestigationGraphEdge[];
  ontology_version: number;
}

// ── API Calls ─────────────────────────────────────────────────────────────────

export function getInvestigationPosture(): Promise<InvestigationPosture> {
  return api.get<InvestigationPosture>("/api/v1/investigations/posture");
}

export function listInvestigations(params?: {
  status?: string;
  severity?: string;
  limit?: number;
  offset?: number;
}): Promise<InvestigationCase[]> {
  const q = new URLSearchParams();
  if (params?.status) q.set("status", params.status);
  if (params?.severity) q.set("severity", params.severity);
  if (params?.limit != null) q.set("limit", String(params.limit));
  if (params?.offset != null) q.set("offset", String(params.offset));
  const qs = q.toString();
  return api.get<InvestigationCase[]>(`/api/v1/investigations${qs ? `?${qs}` : ""}`);
}

export function getInvestigation(caseId: string): Promise<InvestigationCase> {
  return api.get<InvestigationCase>(`/api/v1/investigations/${caseId}`);
}

export function getInvestigationTimeline(
  caseId: string,
  limit = 200
): Promise<InvestigationTimelineEvent[]> {
  return api.get<InvestigationTimelineEvent[]>(
    `/api/v1/investigations/${caseId}/timeline?limit=${limit}`
  );
}

export function getInvestigationEvidence(
  caseId: string,
  sourceDomain?: string
): Promise<EvidenceLink[]> {
  const q = sourceDomain ? `?source_domain=${encodeURIComponent(sourceDomain)}` : "";
  return api.get<EvidenceLink[]>(`/api/v1/investigations/${caseId}/evidence${q}`);
}

export function getInvestigationGraph(caseId: string): Promise<InvestigationGraph> {
  return api.get<InvestigationGraph>(`/api/v1/investigations/${caseId}/graph`);
}

export function acknowledgeInvestigation(caseId: string): Promise<{ case_id: string; status: string }> {
  return api.post(`/api/v1/investigations/${caseId}/acknowledge`, {});
}

export function startInvestigation(caseId: string): Promise<{ case_id: string; status: string }> {
  return api.post(`/api/v1/investigations/${caseId}/start-investigation`, {});
}

export function resolveInvestigation(
  caseId: string,
  resolution_reason: string,
  notes: string
): Promise<{ case_id: string; status: string }> {
  return api.post(`/api/v1/investigations/${caseId}/resolve`, { resolution_reason, notes });
}

// ── Display helpers ───────────────────────────────────────────────────────────

export function severityColor(severity: string): string {
  switch (severity) {
    case "CRITICAL": return "border-red-700 bg-red-900/30 text-red-300";
    case "HIGH": return "border-orange-700 bg-orange-900/30 text-orange-300";
    case "MEDIUM": return "border-yellow-700 bg-yellow-900/30 text-yellow-300";
    case "LOW": return "border-gray-600 bg-gray-800 text-gray-300";
    default: return "border-gray-600 bg-gray-800 text-gray-300";
  }
}

export function confidenceColor(confidence: string): string {
  switch (confidence) {
    case "VERY_HIGH": return "border-purple-700 bg-purple-900/30 text-purple-300";
    case "HIGH": return "border-blue-700 bg-blue-900/30 text-blue-300";
    case "MEDIUM": return "border-cyan-700 bg-cyan-900/30 text-cyan-300";
    case "LOW": return "border-gray-600 bg-gray-800 text-gray-400";
    default: return "border-gray-600 bg-gray-800 text-gray-400";
  }
}

export function statusColor(status: string): string {
  switch (status) {
    case "OPEN": return "border-red-700 bg-red-900/20 text-red-300";
    case "ACKNOWLEDGED": return "border-yellow-700 bg-yellow-900/20 text-yellow-300";
    case "INVESTIGATING": return "border-blue-700 bg-blue-900/20 text-blue-300";
    case "RESOLVED": return "border-green-700 bg-green-900/20 text-green-300";
    default: return "border-gray-600 bg-gray-800 text-gray-300";
  }
}

export function domainColor(domain: string): string {
  switch (domain) {
    case "behavior": return "border-purple-700 bg-purple-900/20 text-purple-300";
    case "ddos": return "border-red-700 bg-red-900/20 text-red-300";
    case "threat_intel": return "border-orange-700 bg-orange-900/20 text-orange-300";
    default: return "border-gray-600 bg-gray-800 text-gray-300";
  }
}

export function domainLabel(domain: string): string {
  switch (domain) {
    case "behavior": return "Behavior";
    case "ddos": return "DDoS";
    case "threat_intel": return "Threat Intel";
    default: return domain;
  }
}

export function formatTs(iso: string | null): string {
  if (!iso) return "—";
  return new Date(iso).toLocaleString(undefined, {
    year: "numeric", month: "short", day: "numeric",
    hour: "2-digit", minute: "2-digit",
  });
}
